codec-http2/src/main/java/io/netty/handler/codec/http2/Http2GoAwayMessage.java

```diff
@@ -0,0 +1,166 @@
+ * under the License.
+ */
+package io.netty.handler.codec.http2;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufHolder;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.util.IllegalReferenceCountException;
+
+public class Http2GoAwayMessage extends Http2Message implements ByteBufHolder {
+    private final long errorCode;
+    private final ByteBuf content;
+    private int extraStreamIds;
+
+    /**
+     * Equivalent to {@code new Http2GoAwayMessage(error.code())}.
+     *
+     * @param error non-{@code null} reason for the go away
+     */
+    public Http2GoAwayMessage(Http2Error error) {
+        this(error.code());
+    }
+
+    /**
+     * Equivalent to {@code new Http2GoAwayMessage(content, Unpooled.buffer(0, 0))}.
+     *
+     * @param error reason for the go away
+     */
+    public Http2GoAwayMessage(long errorCode) {
+        this(errorCode, Unpooled.buffer(0, 0));
+    }
+
+    /**
+     * Equivalent to {@code new Http2GoAwayMessage(error.code(), content)}.
+     *
+     * @param error non-{@code null} reason for the go away
+     * @param content non-{@code null} debug data
+     */
+    public Http2GoAwayMessage(Http2Error error, ByteBuf content) {
+        this(error.code(), content);
+    }
+
+    /**
+     * Construct a new GOAWAY message.
+     *
+     * @param error reason for the go away
+     * @param content non-{@code null} debug data
+     */
+    public Http2GoAwayMessage(long errorCode, ByteBuf content) {
+        this.errorCode = errorCode;
+        this.content = checkNotNull(content, "content");
+    }
+
+    public long errorCode() {
+        return errorCode;
+    }
+
+    /**
+     * The number of IDs to reserve for the receiver to use while GOAWAY is in transit. This allows
+     * for new streams currently en route to still be created, up to a point, which allows for very
+     * graceful shutdown of both sides.
+     */
+    public int extraStreamIds() {
+        return extraStreamIds;
+    }
+
+    /**
+     * Sets the number of IDs to reserve for the receiver to use while GOAWAY is in transit.
+     *
+     * @see #extraStreamIds
+     * @return {@code this}
+     */
+    public Http2GoAwayMessage setExtraStreamIds(int extraStreamIds) {
+        if (extraStreamIds < 0) {
+            throw new IllegalArgumentException("extraStreamIds must be non-negative");
+        }
+        this.extraStreamIds = extraStreamIds;
+        return this;
+    }
+
+    @Override
+    void process(ChannelHandlerContext ctx, Http2ConnectionHandler http2Handler,
+            ChannelPromise promise) {
+        int lastStreamId = http2Handler.connection().remote().lastStreamCreated() + extraStreamIds;
+        http2Handler.goAway(ctx, lastStreamId, errorCode, content.retain(), promise);
```

Comment:
That is what is happening now (assuming same eventLoop between parent and child). But now that you are bringing my attention to it, that is wrong. We should be doing a retain before writing to the parent channel and make sure to do a release after calling `process()`.

A small detail is that `Http2GoAwayMessage` is only on the parent channel, since it isn't stream-based. So `ChildChannel` doesn't come into play for this one. And the need for release after calling `process()` is even more apparent.